### ISSUE TITLE
zfs receive fails with space quota exceeded for compressed dataset

### ISSUE THREAD

**Issue by lazy404** [openzfs/zfs#4492]
sender 0.6.5.4
receiver 0.6.5.5

on sender

```
wwwpool/xxx_www  compression           lz4
wwwpool/xxx_www  compressratio         1.44x
wwwpool/xxx_www  usedbydataset         94,8M
wwwpool/xxx_www  refquota              97,7M
wwwpool/xxx_www  logicalreferenced     123M

```

```
# zfs send -peD wwwpool/xxx_www@xxx|ssh x.x.x.x zfs receive -vd wwwpool
receiving full stream of wwwpool/xxx_www@xxx into wwwpool/xxx_www@xxx
cannot receive new filesystem stream: destination wwwpool space quota exceeded
```

output produced by zfs send is ~120MB

```
root@xxx:~# zfs send -pv wwwpool/xxx@xxx > test && ls -l test
send from @ to wwwpool/xxx@xxx estimated size is 123M
total estimated size is 123M
TIME        SENT   SNAPSHOT
-rw-r--r-- 1 root root 133232672 kwi  6 10:36 test

root@xxx:~# zfs send -pev wwwpool/xxx@xxx > test && ls -l test
send from @ to wwwpool/xxx@xxx estimated size is 123M
total estimated size is 123M
TIME        SENT   SNAPSHOT
-rw-r--r-- 1 root root 133015944 kwi  6 10:36 test

root@xxx:~# zfs send -peDv wwwpool/xxx@xxx > test && ls -l test
send from @ to wwwpool/xxx@xxx estimated size is 123M
total estimated size is 123M
TIME        SENT   SNAPSHOT
10:36:49   96,7M   wwwpool/xxx@xxx
-rw-r--r-- 1 root root 125021064 kwi  6 10:36 test
```

ther's a lot of free space on destination pool and lz4 compression is also enabled there
for dataset's not exceeding refquota when uncompressed send/recv succedes without any problems

**Comment by gmelikov** [comment]
Need some logs and dataset infomation. I'm closing issue for now, you can reopen it if it's actual

**Comment by williamstein** [comment]
This is the same as https://www.illumos.org/issues/4986 and is definitely a real issue still..

**Comment by zmielna** [comment]
Bump, I'm hit by this too, CentOS Linux release 7.3.1611 on both source and target.

**Comment by behlendorf** [comment]
Reopening.  @zmielna @williamstein can you post the zfs version you were using.

**Comment by kernelOfTruth** [comment]
Referencing commits:

https://github.com/zfsonlinux/zfs/commit/671c93546cc0471ff3b4039e82d6f2a77de6d413 OpenZFS 4986 - receiving replication stream fails if any snapshot exc…
…eeds refquota

https://github.com/zfsonlinux/zfs/commit/8c62a0d0f39c450d1fdb001b8073193f89611f56 OpenZFS 6562 - Refquota on receive doesn't account for overage

(both in-tree since 0.7.0-rc1)

**Comment by zmielna** [comment]
Certainly, it's kmod zfs package in my case and this here is zfs version on source machine:

````# modinfo zfs|head
filename:       /lib/modules/3.10.0-514.6.2.el7.x86_64/weak-updates/zfs.ko
version:        0.6.5.9-1
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
rhelversion:    7.3
srcversion:     42C4AB70887EA26A9970936
depends:        spl,znvpair,zcommon,zunicode,zavl
vermagic:       3.10.0-514.6.1.el7.x86_64 SMP mod_unload modversions 
```

and this is target of zfs send

```# modinfo zfs|head
filename:       /lib/modules/3.10.0-514.26.2.el7.x86_64/extra/zfs/zfs/zfs.ko
version:        0.7.1-1
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
rhelversion:    7.3
srcversion:     C1E2CDED3042134D600CF4E
depends:        spl,znvpair,zcommon,zunicode,zavl,icp
vermagic:       3.10.0-514.26.2.el7.x86_64 SMP mod_unload modversions 
````

**Comment by williamstein** [comment]
I'm using
```
root@kucalc-prod-node-t6tt:/home/cocalc# modinfo zfs|head
filename:       /lib/modules/4.10.0-32-generic/kernel/zfs/zfs/zfs.ko
version:        0.6.5.9-2
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
srcversion:     42C4AB70887EA26A9970936
depends:        spl,znvpair,zcommon,zunicode,zavl
vermagic:       4.10.0-32-generic SMP mod_unload
parm:           zvol_inhibit_dev:Do not create zvol device nodes (uint)
parm:           zvol_major:Major number for zvol device (uint)
root@kucalc-prod-node-t6tt:/home/cocalc# cat /etc/issue
Ubuntu 17.04 \n \l
```

And -- in case you want to TRULY CRINGE -- here is the actual Python code I'm using in production (on cocalc.com) to workaround this bug (the code recvs a bunch of streams that are gcsfuse mounted from google cloud storage buckets):

```
def recv_streams(config):
    ensure_gcs_bucket_is_mounted(config)
    path = streams_path(config)
    if not os.path.exists(path):
        return  # no streams to recv
    streams = [x for x in os.listdir(path) if x.endswith('.lz4')]
    streams.sort()
    if len(streams) == 0:
        return  # no streams
    cmd("zfs destroy -r {pool_name}; zfs set refquota=none {pool_name}".format(**config))
    # Horrible hack -- when creating the pool a potentially too-small
    # refquota gets set, which may prevent import due to this
    # ZFS bug: https://www.illumos.org/issues/4986
    # This only happens if the ZFS stream is sufficiently large that this
    # recv will take some time.  So we set the refquota back to none **after**
    # the recv starts.
    def f(n):
        if n == 0:
            cmd("sleep 1; zfs set refquota=none {pool_name}".format(**config))
        elif n == 1:
            try:
                c = "lz4 -d {stream} | zfs recv {pool_name} -F".format(stream=os.path.join(path, streams[0]), **config)
                cmd(c)
            except Exception as err:
                # Hack didn't work (I never saw this) - might as well just try
                # again to import; for some reason it works the second time.
                # Clearly this SUCKS.  Sorry.
                if 'space quota exceeded' in str(err):
                    cmd("zfs set refquota=none {pool_name} && zfs destroy -r {pool_name}".format(**config))
                    cmd(c)
    thread_map(f, [0,1])
    # Now receive the rest.
    for s in streams[1:]:
        cmd("lz4 -d {stream} | zfs recv {pool_name}".format(
                stream=os.path.join(path, s), **config))
```

(and thank you @behlendorf for all you do on zfsonlinux; I'm a huge fan.)

**Comment by kernelOfTruth** [comment]
Referencing further:

https://github.com/zfsonlinux/zfs/commit/fcff0f35bd522076bdda7491c88a91cc0aa531a3 Illumos 5960, 5925
5960 zfs recv should prefetch indirect blocks
5925 zfs receive -o origin=

(also included in-tree since 0.7.0-rc1)
